### PR TITLE
Manage dependencies

### LINKED ISSUE #7: Restart CT scans
# 🐛 Summary #

Certificate Transparency (CT) scans stopped running sometime last June. As a result the section of VS reports that should contain CT logs are instead empty. This issue documents the need to investigate why scans stopped, and will be resolved by changes that get scans to start again.

### PR DESCRIPTION
**Pull request by king-alexander**
#  #

## 🗣 Description ##

Celery 5.0 introduces a new CLI that is not backwards-compatible. To avoid breaking changes, pin the versions of Celery which work with our implementation. 

## 💭 Motivation and context ##

This change resolves Issue #7 and restores our ability to comply with [Emergency Directive 19-01](https://www.cisa.gov/emergency-directive-19-01).

## 🧪 Testing ##

I followed the instructions in `README.md` to run system and unit tests with `docker compose -f docker-compose-dev.yml run test`. All passed. Then I started a bash shell with `docker compose -f docker-compose-dev.yml run bash` and ran admiral on the last 1000 federal domains by alphabetical order with `./examples/load_certs.py -v --skipto==smokeybear.gov`. 

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.

### CODE REVIEW

File: setup.py
Comment (jsf9k) [inline setup.py]: I think this is more accurately captures the version pinning we want to do:
```suggestion
        "celery >=4.3.0,<5",
```

You should probably also add a comment to `setup.py` just above this line explaining why we want to avoid `celery` 5 and above.

The following quote is from [this source](https://packaging.python.org/en/latest/discussions/install-requires-vs-requirements/):

> It is not considered best practice to use `install_requires` to pin dependencies to specific versions, or to specify sub-dependencies (i.e. dependencies of your dependencies). This is overly-restrictive, and prevents the user from gaining the benefit of dependency upgrades.

File: Dockerfile-admiral
Comment (jsf9k) [inline Dockerfile-admiral]: I don't think this belongs here.  Note that `mongomock` [is already included as part of the `test` extra listed in `setup.py`](https://github.com/cisagov/admiral/blob/88e530496c4a06b177a58f6927281420a8c7b801/setup.py#L87).

If you are attempting to run the tests from inside the Docker container then you should instead temporarily modify _your local copy_ of `Dockerfile-admiral` so that line 34 becomes:
```python
pip install -e .[test]
```

This change should not be committed to the repository, though.

**Review by jsf9k — CHANGES_REQUESTED** [review]
I mentioned a couple of things that should be done a little differently, if I correctly understand the reasoning behind doing them.  Let me know if I've got it wrong and I'm happy to reconsider.

File: Dockerfile-admiral
Comment (jsf9k) [reply Dockerfile-admiral]: @felddy, can you explain why `pytest` is listed in the line above?  I don't see it as necessary for the purposes of running the admiral either.  `pytest` is included as part of the `test` extra, just like `mongomock`.

**Comment by jsf9k** [comment]
@mcdonnnj, any idea why the lint and test actions aren't kicking off?

File: Dockerfile-admiral
Comment (mcdonnnj) [reply Dockerfile-admiral]: It looks like it's using the `INSTALL_IPYTHON` environment variable to set up a development Docker image per https://github.com/cisagov/admiral/commit/147f9a8dbf9421f83efb7c7034ed2250da786d8f.

Comment (jsf9k) [reply Dockerfile-admiral]: I see...but shouldn't it instead just `pip install -e .[test]` (or `pip install -e .[dev]`) vice `pip install -e .`, depending on the value of `INSTALL_IPYTHON`?  That seems simpler than re-listing all the dev dependencies in the `Dockerfile`.

**Comment by jsf9k** [comment]
@king-alexander, before this PR gets merged we should rebase to remove the first two commits.  There is no reason they should clutter up the commit history.

File: setup.py
Comment (jsf9k) [reply setup.py]: Fixed in e97e46a.

@king-alexander - note that you could have just clicked the "Commit suggestion" button above and created a suitable commit message.  That's why I did it that way.

**Comment by king-alexander** [comment]
Okay, rebasing took me a few tries but I figured it out! Thank you @jsf9k for the comments. I think linting and actions aren't kicking off yet because I marked the PR as a draft. I'll make it ready for review now.

File: setup.py
Comment (king-alexander) [reply setup.py]: Thanks! Yes I saw but I wanted to include the versioning comment you suggested in the same commit.

File: Dockerfile-admiral
Comment (mcdonnnj) [reply Dockerfile-admiral]: That's probably a saner approach (preferring `--editable` over `-e`).

Comment (jsf9k) [reply Dockerfile-admiral]: Please see commit a73bde9.

**Review by jsf9k — COMMENTED** [review]
These changes look good to me, but I'd like to see the GH Actions checks in the workflow passing before I approve.  @mcdonnnj, do you know why they aren't running?

**Comment by dav3r** [comment]
Is the goal of this PR to get our CT log scanner working again or simply to ensure that the celery piece runs?  If it's the former, I'd like to see some attestation (e.g. in the "Testing" section of this PR) that the scanner is able to be run successfully.

Also, doesn't this PR contradict the goal specified in issue #6 to move to Celery 5.0?

**Comment by jsf9k** [comment]
> Is the goal of this PR to get our CT log scanner working again or simply to ensure that the celery piece runs? If it's the former, I'd like to see some attestation (e.g. in the "Testing" section of this PR) that the scanner is able to be run successfully.
> 
> Also, doesn't this PR contradict the goal specified in issue #6 to move to Celery 5.0?

I believe this PR represents just the baby step of getting the code to pass its unit tests again.  I don't think this PR is intended to address #6.

If this PR has high ambitions than that then I agree that it needs more testing than simply passing the unit tests.  Please chime in @king-alexander.

**Comment by king-alexander** [comment]
That's correct @jsf9k. I intended this PR as a baby step. Now that I can get unit tests to pass, I'll try to get the scanner to run again. I opened #6 to document an improvement we will eventually want to make. But I don't think it's work the effort yet.

**Comment by dav3r** [comment]
> That's correct @jsf9k. I intended this PR as a baby step. Now that I can get unit tests to pass, I'll try to get the scanner to run again. I opened #6 to document an improvement we will eventually want to make. But I don't think it's work the effort yet.

Makes sense, thanks for the clarification.  Can you please create an issue that describes getting the scanning running again?

**Review by dav3r — APPROVED** [review]
The code changes look good, but I agree with @jsf9k that we need to figure out why the GitHub Actions (lint, test, build) aren't running here.

**Comment by dav3r** [comment]
> > That's correct @jsf9k. I intended this PR as a baby step. Now that I can get unit tests to pass, I'll try to get the scanner to run again. I opened #6 to document an improvement we will eventually want to make. But I don't think it's work the effort yet.
> 
> Makes sense, thanks for the clarification. Can you please create an issue that describes getting the scanning running again?

@king-alexander Just closing the loop here- it looks like you created #7 to address my concern above, correct?

**Comment by king-alexander** [comment]
> > > That's correct @jsf9k. I intended this PR as a baby step. Now that I can get unit tests to pass, I'll try to get the scanner to run again. I opened #6 to document an improvement we will eventually want to make. But I don't think it's work the effort yet.
> > 
> > 
> > Makes sense, thanks for the clarification. Can you please create an issue that describes getting the scanning running again?
> 
> @king-alexander Just closing the loop here- it looks like you created #7 to address my concern above, correct?

That's correct!

**Comment by jsf9k** [comment]
@king-alexander, @mcdonnnj has some upstream updates (from [cisagov/skeleton-python-library](https://github.com/cisagov/skeleton-python-library)) that should make the GH Actions workflows work again.  When the time comes we'll rebase this branch on top of those changes and then we can look at merging.

**Comment by king-alexander** [comment]
> @king-alexander, @mcdonnnj has some upstream updates (from [cisagov/skeleton-python-library](https://github.com/cisagov/skeleton-python-library)) that should make the GH Actions workflows work again. When the time comes we'll rebase this branch on top of those changes and then we can look at merging.

Sounds good 👍

**Comment by king-alexander** [comment]
@jsf9k @mcdonnnj Are y'all ready for me to rebase these changes onto #8?

**Comment by jsf9k** [comment]
> @jsf9k @mcdonnnj Are y'all ready for me to rebase these changes onto #8?

No, let's wait until #8 gets merged.

**Review by mcdonnnj — CHANGES_REQUESTED** [review]
@king-alexander Once you've rebased on `develop` now that #8 is merged please use the `bump_version.sh` script to bump the minor version in this project.

**Comment by jsf9k** [comment]
@king-alexander, did you only rebase and force push?  I thought I had [at least one commit in here](https://github.com/cisagov/admiral/commit/a73bde9790fe6143b0c54a2f388de082553ea935) and now I have none.  I just want to make sure that none of the changes we made before got dropped.

**Comment by king-alexander** [comment]
I believe I have a new change that makes admiral compatible with the latest version of Celery (v5.0 Singularity). I am marking this PR as a draft again to do more comprehensive testing.

**Comment by king-alexander** [comment]
The new change isn't ready yet, so I am reopening.

**Comment by jsf9k** [comment]
> The new change isn't ready yet, so I am reopening.

I was thinking that I'd prefer they be separate PRs anyway.

**Review by jsf9k — APPROVED** [review]
The only difference I see between this code and the code in @mcdonnnj's testing branch is the version number, so I think this PR is ready to be merged.

**Review by mcdonnnj — APPROVED** [review]

**Pull request merged**
